Add WebSockets conceptual doc (#32058)
dotnet/docs: docs/fundamentals/networking/websockets.md

---
title: WebSockets support in .NET
description: Learn how to web sockets with the ClientWebSockets in .NET.
author: greenEkatherine
ms.author: esokolov
ms.date: 10/27/2022
---

# WebSockets support in .NET

The WebSocket Protocol enables two-way communication between a client to a remote host that has opted-in to communications from that code. The <xref:System.Net.WebSockets.ClientWebSocket?displayProperty=fullName> exposes the ability to establish WebSocket connection via an opening handshake: it is created and sent by `ConnectAsync` method.

> ## HTTP/2 vs HTTP/1.1 WebSockets

Reviewer: ```suggestion
## Differences in HTTP/1.1 and HTTP/2 WebSockets
```

> WebSockets over HTTP/1.1 uses a single TCP connection, therefore it is managed by connection-wide headers (see [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455)). Here is an example of how to establish WebSocket over HTTP/1.1:

Reviewer: ```suggestion
WebSockets over HTTP/1.1 uses a single TCP connection, therefore it is managed by connection-wide headers, for more information, see [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455). Consider the following example of how to establish WebSocket over HTTP/1.1:
```

```c#
ClientWebSocket ws = new();
ws.ConnectAsync(uri, cancellationToken);
```

> A different approach must be taken with HTTP/2 due to its multiplexing nature: here, WebSockets are established per stream (see [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441)). With HTTP/2 it is possible to use one connection for multiple web socket streams together with ordinary HTTP streams and extends HTTP/2's more efficient use of the network to WebSockets. There is a special overload of `ConnectAsync` which accepts `HttpMessageInvoker` to allow reusing existing pooled connections:

Reviewer: ```suggestion
A different approach must be taken with HTTP/2 due to its multiplexing nature. WebSockets are established per stream, for more information, see [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441). With HTTP/2 it is possible to use one connection for multiple web socket streams together with ordinary HTTP streams and extend HTTP/2's more efficient use of the network to WebSockets. There is a special overload of <xref:System.Net.WebSockets.ClientWebSocket.ConnectAsync(System.Uri,System.Net.Http.HttpMessageInvoker,System.Threading.CancellationToken)> which accepts an <xref:System.Net.Http.HttpMessageInvoker> to allow reusing existing pooled connections:
```

```c#
var handler = new SocketsHttpHandler();
ClientWebSocket ws = new();

> ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);

Author: A user should dispose

```

## How to set up HTTP version and policy

By default, `ClientWebSocket` uses HTTP/1.1 to send an opening handshake and allows downgrade. It can be changed before calling `ConnectAsync`:

```c#

> ClientWebSocket ws = new();

Reviewer: ```suggestion
SocketsHttpHandler handler = new();
ClientWebSocket ws = new();
```

ws.Options.HttpVersion = HttpVersion.Version20;
ws.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher;

> ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);

Reviewer: ```suggestion
await ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);
```

```

> ## Incompatible options

Reviewer 2: Can we please have an example on how to set the options on the invoker instead of the options here? It could be taken from https://github.com/dotnet/docs/blob/8239d82966dd63f3d405fc808cfe5794fb784bd0/docs/core/compatibility/networking/7.0/connectasync-argumentexception.md#recommended-action

`ClientWebSocket` has properties <xref:System.Net.WebSockets.ClientWebSocketOptions?displayProperty=fullName> that the user can set up before the connection is established. However, when `HttpMessageInvoker` is provided, it also has these properties. To avoid ambiguity, in that case, properties should be set on `HttpMessageInvoker`, and `ClientWebSocketOptions` should have default values. Otherwise, if `ClientWebSocketOptions` are changed, overload of `ConnectAsync` will throw an <xref:System.ArgumentException>.

> ## Compression

Reviewer 2: Can we have an example here please? It could be taken from https://devblogs.microsoft.com/dotnet/announcing-net-6-preview-5/#libraries-websocket-compression

Reviewer 2: example on how to turn off compression for specific messages would be also very useful (so it is not only said in text here)

Author: > example on how to turn off compression for specific messages would be also very useful (so it is not only said in text here)

I cannot find how to do it

Reviewer 2: Huh... Turns out we have this option available only for `WebSocket`, not `ClientWebSocket` (see `WebSocketMessageFlags` in [WebSocket.SendAsync](https://learn.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket.sendasync?view=net-6.0#system-net-websockets-websocket-sendasync(system-readonlymemory((system-byte))-system-net-websockets-websocketmessagetype-system-net-websockets-websocketmessageflags-system-threading-cancellationtoken))) 🤦‍♀️🤦‍♀️🤦‍♀️ We should expose this API on `ClientWebSocket` as well. Will you be able to create an issue for that pls? Or I can do that.

Author: Ok, I will create an issue

Author: https://github.com/dotnet/runtime/issues/77783, did I miss anything? If not, I will update it with API proposal a little bit later

> WebSocket supports per-message deflate as defined in [RFC 7692](https://tools.ietf.org/html/rfc7692#section-7). It is controlled by <xref:System.Net.WebSockets.ClientWebSocketOptions.DangerousDeflateOptions?displayProperty=fullName>. When present, the options are sent to the server during the handshake phase. If the server supports per-message-deflate and the options are accepted, the `WebSocket` instance will be created with compression enabled by default for all messages.

Author: Yes, one RFC for web sockets over HTTP/1.1, another one for web sockets over HTTP/2, and the third one for compression.

---
title: WebSockets support in .NET
description: Learn how to web sockets with the ClientWebSockets in .NET.
author: greenEkatherine
ms.author: esokolov
ms.date: 10/27/2022
---

# WebSockets support in .NET

The WebSocket protocol enables two-way communication between a client and a remote host. The <xref:System.Net.WebSockets.ClientWebSocket?displayProperty=fullName> exposes the ability to establish WebSocket connection via an opening handshake, it is created and sent by the `ConnectAsync` method.

## Differences in HTTP/1.1 and HTTP/2 WebSockets

WebSockets over HTTP/1.1 uses a single TCP connection, therefore it is managed by connection-wide headers, for more information, see [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455). Consider the following example of how to establish WebSocket over HTTP/1.1:

```c#

> ClientWebSocket ws = new();

Reviewer: ```suggestion
using ClientWebSocket ws = new();
```

await ws.ConnectAsync(uri, cancellationToken);
```

A different approach must be taken with HTTP/2 due to its multiplexing nature. WebSockets are established per stream, for more information, see [RFC 8441](https://www.rfc-editor.org/rfc/rfc8441). With HTTP/2 it is possible to use one connection for multiple web socket streams together with ordinary HTTP streams and extend HTTP/2's more efficient use of the network to WebSockets. There is a special overload of <xref:System.Net.WebSockets.ClientWebSocket.ConnectAsync(System.Uri,System.Net.Http.HttpMessageInvoker,System.Threading.CancellationToken)> which accepts an <xref:System.Net.Http.HttpMessageInvoker> to allow reusing existing pooled connections:

```c#

> SocketsHttpHandler handler = new();
> ClientWebSocket ws = new();

Reviewer: ```suggestion
using SocketsHttpHandler handler = new();
using ClientWebSocket ws = new();
```

await ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);
```

## Set up HTTP version and policy

By default, `ClientWebSocket` uses HTTP/1.1 to send an opening handshake and allows downgrade. In .NET 7 web sockets over HTTP/2 are available. It can be changed before calling `ConnectAsync`:

```c#

> SocketsHttpHandler handler = new();
> ClientWebSocket ws = new();

Reviewer: ```suggestion
using SocketsHttpHandler handler = new();
using ClientWebSocket ws = new();
```

ws.Options.HttpVersion = HttpVersion.Version20;
ws.Options.HttpVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher;

await ws.ConnectAsync(uri, new HttpMessageInvoker(handler), cancellationToken);
```

## Incompatible options

`ClientWebSocket` has properties <xref:System.Net.WebSockets.ClientWebSocketOptions?displayProperty=fullName> that the user can set up before the connection is established. However, when `HttpMessageInvoker` is provided, it also has these properties. To avoid ambiguity, in that case, properties should be set on `HttpMessageInvoker`, and `ClientWebSocketOptions` should have default values. Otherwise, if `ClientWebSocketOptions` are changed, overload of `ConnectAsync` will throw an <xref:System.ArgumentException>.

```c#

> var handler = new HttpClientHandler();
> handler.CookieContainer = cookies;
> handler.UseCookies = cookies != null;
> handler.ServerCertificateCustomValidationCallback = remoteCertificateValidationCallback;
> handler.Credentials = useDefaultCredentials ?
>     CredentialCache.DefaultCredentials :
>     credentials;
> if (proxy == null)
> {
>     handler.UseProxy = false;
> }
> else
> {
>     handler.Proxy = proxy;
> }
> if (clientCertificates?.Count > 0)
> {
>     handler.ClientCertificates.AddRange(clientCertificates);
> }
> var invoker = new HttpMessageInvoker(handler);
> 
> var cws = new ClientWebSocket();
> await cws.ConnectAsync(uri, invoker, cancellationToken);

Reviewer: ```suggestion
using HttpClientHandler handler = new()
{
    CookieContainer = cookies;
    UseCookies = cookies != null;
    ServerCertificateCustomValidationCallback = remoteCertificateValidationCallback;
    Credentials = useDefaultCredentials
        ? CredentialCache.DefaultCredentials
        : credentials;
};
if (proxy is null)
{
    handler.UseProxy = false;
}
else
{
    handler.Proxy = proxy;
}
if (clientCertificates?.Count > 0)
{
    handler.ClientCertificates.AddRange(clientCertificates);
}
HttpMessageInvoker invoker = new(handler);
using ClientWebSocket cws = new();
await cws.ConnectAsync(uri, invoker, cancellationToken);
```

```

## Compression

The WebSocket protocol supports per-message deflate as defined in [RFC 7692](https://tools.ietf.org/html/rfc7692#section-7). It is controlled by <xref:System.Net.WebSockets.ClientWebSocketOptions.DangerousDeflateOptions?displayProperty=fullName>. When present, the options are sent to the server during the handshake phase. If the server supports per-message-deflate and the options are accepted, the `ClientWebSocket` instance will be created with compression enabled by default for all messages.

```c#

> var cws = new ClientWebSocket();
> cws.Options.DangerousDeflateOptions = new WebSocketDeflateOptions()
> {
>     ClientMaxWindowBits = 10,
>     ServerMaxWindowBits = 10
> };

Reviewer: ```suggestion
using ClientWebSocket ws = new()
{
    Options =
    {
        DangerousDeflateOptions = new WebSocketDeflateOptions()
        {
            ClientMaxWindowBits = 10,
            ServerMaxWindowBits = 10
        }
    }
};
```
